keps/262-ConfigurableFailurePolicy/README.md

```diff
@@ -0,0 +1,608 @@
+will be restarting numerous times along the road to completing this very long training job, the workers
+should be able to restart an unlimited number of times.
+
+When the parameter service in the `parameter-server` replicated job fails, I want the JobSet
+to restart but only up to 3 times, as the job 
+
+**Example Failure Policy Configuration for this use case**:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: different-policies-for-different-replicated-jobs
+spec:
+  # Failure Policy configured to restart the JobSet upon any failure, up to 10 times.
+  # Otherwise, restart up to 10 times. This rule applies to any failure type for all
+  # replicated jobs.
+  failurePolicy:
+    rules:
+    - action: RestartJobSetAndIgnoreMaxRestarts
+      targetReplicatedJobs:
+      - workers
+    - action: RestartJobSet
+      targetReplicatedJobs:
+      - parameter-server
+  replicatedJobs:
+  - name: workers
+    replicas: 5
+    template:
+      spec:
+        parallelism: 3000
+        completions: 3000
+        backoffLimit: 0
+        template:
+          spec:
+            restartPolicy: Never
+            containers:
+            - name: main
+              image: bash:latest
+              image: python:3.8
+              command: 
+              - |
+                python3 train.py
+```
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Proposed Failure Policy API
+
+```go
+
+// FailurePolicyAction defines the action the JobSet controller will take for
+// a given FailurePolicyRule.
+type FailurePolicyAction string
+
+const (
+  // Fail the JobSet immediately, regardless of maxRestarts.
+  FailJobSet FailurePolicyAction = "FailJobSet"
+
+  // Restart the JobSet if the number of restart attempts is less than MaxRestarts.
+  // Otherwise, fail the JobSet.
+  RestartJobSet FailurePolicyAction = "RestartJobSet"
+
+  // Don't count the failure against maxRestarts.
+  RestartJobSetAndIgnoreMaxRestarts FailurePolicyAction = "RestartJobSetAndIgnoreMaxRestarts"
+)
+
+// FailurePolicyRule defines a FailurePolicyAction to be executed if a child job
+// fails due to a reason listed in OnJobFailureReasons.
+type FailurePolicyRule struct {
+  // The action to take if the rule is matched.
+  // +kubebuilder:validation:Enum:=FailJobSet;RestartJobSetAndIgnoreMaxRestarts;FailJob;RestartJob
```

Comment:
note for implementation: remove FailJob;RestartJob from kubebuilder marker. We'll add them back after upstream changes are in place to support these use cases.